velox/type/DecimalUtil.h

```diff
@@ -203,32 +204,57 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
-  /// Rescales a double value to decimal value of given precision and scale. The
-  /// output is rescaled value of int128_t or int64_t type. Returns error status
-  /// if fails.
-  template <typename TOutput>
-  inline static Status
-  rescaleDouble(double value, int precision, int scale, TOutput& output) {
+  /// Rescales a floating point value to decimal value of given precision and
+  /// scale. The output is rescaled value of int128_t or int64_t type. Returns
+  /// error status if fails.
+  template <typename TIntput, typename TOutput>
+  inline static Status rescaleFloatingPoint(
+      TIntput value,
+      int precision,
+      int scale,
+      TOutput& output) {
     if (!std::isfinite(value)) {
       return Status::UserError("The input value should be finite.");
     }
 
-    long double rounded;
-    // A double provides 16(±1) decimal digits, so at least 15 digits are
-    // precise.
-    if (scale > 15) {
-      // Convert value to long double type, as double * int128_t returns
-      // int128_t and fractional digits are lost. No need to consider 'toValue'
-      // becoming infinite as DOUBLE_MAX * 10^38 < LONG_DOUBLE_MAX.
-      const auto toValue = (long double)value * DecimalUtil::kPowersOfTen[15];
-      rounded = std::round(toValue) * DecimalUtil::kPowersOfTen[scale - 15];
+    uint8_t digits;
+    if constexpr (std::is_same_v<TIntput, float>) {
+      // A float provides between 6 and 7 decimal digits, so at least 6 digits
+      // are precise.
+      digits = 6;
+    } else {
+      // A double provides from 15 to 17 decimal digits, so at least 15 digits
+      // are precise.
+      digits = 15;
+      if (value <= std::numeric_limits<int128_t>::min() ||
+          value >= std::numeric_limits<int128_t>::max()) {
+        return Status::UserError("Result overflows.");
```

Comment:
Would users understand this error? Should we clarify a bit and include the value in the message?

Here is what Presto reports:

```
presto:feed> select cast(pow(10, 38) - pow(10, 21) as decimal(38, 0));
Query 20240229_152537_16711_hedpg failed: Cannot cast DOUBLE '1.0E38' to DECIMAL(38, 0)
```

Reply:
The method `makeBadCastException` at [apply](https://github.com/facebookincubator/velox/pull/8575/files#diff-872249bbe84c54f0119d4568444009b9b850ec34b802ebbc0b641bb012fa02baR489) creates an error message of `Cannot cast {} to {}.`, so the full error is like below.
> Cannot cast DOUBLE '-1.7976931348623157E308' to DECIMAL(38, 2). Result overflows.